README
别样连连看

这个是 别样连连看 C 语言 OpenGL 加速的gtk版本。







NEWS
Sample NEWS file for 俄罗斯方块 project.


src/byllk.c

#include <config.h>

#include <stdlib.h>
#include <time.h>
#include <locale.h>
#include <glib/gi18n.h>
#include <gtk/gtk.h>
#include <gtk/gtkgl.h>
#include <GL/glu.h>
#include <GL/glx.h>
#include "byllk.h"

static void on_realize(GtkWidget *widget, gpointer user_data);
static gboolean on_resize(GtkWidget *widget, GdkEventConfigure *event,gpointer user_data);

static gboolean on_click(GtkWidget *widget, GdkEventButton *event, gpointer user_data)
{
	DisplayResource	*displayrc;
	gint	width,height;

	g_object_get(widget,"user-data",&displayrc,NULL);

	gdk_drawable_get_size(widget->window,&width,&height);

	double xsize =(double)width/LINEX;
	double ysize =(double)height/LINEY;

	displayrc->clicked.x = event->x/ xsize;
	displayrc->clicked.y = LINEY - event->y / ysize;

	//好了，检测

	if(displayrc->balls[displayrc->clicked.x][displayrc->clicked.y]>=0)
	{
		displayrc->clicked.x = displayrc->clicked.y = -1;
	}else //点到的是空白的，呵呵
	{
		GdkPoint		balls[4];
		GLint		balls_color[4]={-1,-1,-1,-1};
		//四个方向发散
		balls[0] = displayrc->clicked;
		balls[0].y = displayrc->clicked.y + 1;

		balls[1] = displayrc->clicked;
		balls[1].y = displayrc->clicked.y - 1;

		balls[2] = displayrc->clicked;
		balls[2].x = displayrc->clicked.x - 1;

		balls[3] = displayrc->clicked;
		balls[3].x = displayrc->clicked.x + 1;

		while(  balls[0].y < LINEY && displayrc->balls[balls[0].x][balls[0].y] <0 )
			balls[0].y ++;
		if(balls[0].y < LINEY)
			balls_color[0] = displayrc->balls[balls[0].x][balls[0].y];

		while(  balls[1].y > 0 && displayrc->balls[balls[1].x][balls[1].y] <0 )
			balls[1].y --;
		balls_color[1] = displayrc->balls[balls[1].x][balls[1].y];

		while(  balls[2].x >0  && displayrc->balls[balls[2].x][balls[2].y] <0 )
			balls[2].x --;
		balls_color[2] = displayrc->balls[balls[2].x][balls[2].y];

		while(  balls[3].x < LINEX  && displayrc->balls[balls[3].x][balls[3].y] <0 )
			balls[3].x ++;
		if(balls[3].x < LINEX)
			balls_color[3] = displayrc->balls[balls[3].x][balls[3].y];

		//处理好了，现在，我们需要去掉碰撞的啦
		for(int i=0;i<3;i++)
		{
			if(balls_color[i]<0)continue;
			for(int j=i+1;j<4;j++)
			{
				if(balls_color[j]<0)continue;
				if(balls_color[i] == balls_color[j])
				{
					displayrc->balls[balls[i].x][balls[i].y]=-1;
					displayrc->balls[balls[j].x][balls[j].y]=-1;
				}
			}
		}
	}
	gtk_widget_queue_draw(widget);
	return FALSE;
}

static gboolean on_iconfy(GtkWidget *widget, GdkEventWindowState *event,GtkWidget *draw_area)
{
	if(event->new_window_state & GDK_WINDOW_STATE_ICONIFIED)
	{

	}else
	{

	}
	return FALSE;
}

int main(int argc, char * argv[], char * env[])
{
	GtkWidget* window;
	GtkWidget* draw_area;

	setlocale(LC_ALL,"");
	textdomain(PACKAGE_NAME);

	g_thread_init(NULL);

	gtk_init(&argc,&argv);
	gtk_gl_init(&argc,&argv);

	window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
	draw_area = gtk_drawing_area_new();
	gtk_container_add(GTK_CONTAINER(window),draw_area);

	g_assert(glconfig = gdk_gl_config_new_by_mode(GDK_GL_MODE_RGBA|GDK_GL_MODE_DOUBLE|GDK_GL_MODE_MULTISAMPLE|GDK_GL_MODE_DEPTH));
	g_assert(gtk_widget_set_gl_capability(draw_area,glconfig,NULL,TRUE,GDK_GL_RGBA_TYPE));


	g_signal_connect(window,"destroy",gtk_main_quit,NULL);
	g_signal_connect(draw_area,"expose-event",G_CALLBACK(on_draw),NULL);
	g_signal_connect(draw_area,"configure-event",G_CALLBACK(on_resize),NULL);
	g_signal_connect(draw_area,"realize",G_CALLBACK(on_realize),NULL);
	g_signal_connect(draw_area,"button-press-event",G_CALLBACK(on_click),NULL);

	g_signal_connect(draw_area,"window-state-event",G_CALLBACK(on_iconfy),draw_area);

	gtk_widget_add_events(draw_area,GDK_BUTTON_PRESS_MASK|GDK_EXPOSURE_MASK|GDK_STRUCTURE_MASK);
	gtk_widget_add_events(window,GDK_STRUCTURE_MASK);

	gtk_window_set_default_size(GTK_WINDOW(window),25*32,16*32);

	gtk_widget_show_all(window);

//	g_timeout_add(100,(GSourceFunc)(gtk_widget_queue_draw),draw_area);

	gtk_main();
	return 0;
}

static void build_projection()
{
	glMatrixMode(GL_PROJECTION);

	glLoadIdentity();

	glFrustum(-0.5, 0.5, -0.5, 0.5, 1, 1000);
	glTranslated(0, 0, -2);

	glMatrixMode(GL_MODELVIEW);
}

static void on_realize(GtkWidget *widget, gpointer user_data)
{
	int	x,y;

	DisplayResource	* displayrc = g_new(DisplayResource,1);
	g_object_set(widget,"user-data",displayrc,NULL);

	srand(time(NULL));

	for(x=0;x < LINEX ; x++)
	{
		for(y=0;y<LINEY;y++)
		{
			displayrc->balls[x][y] = rand() % (COLOR_NUM+1) -1;
		}
	}

	GdkGLContext * glcontext;
	GdkGLDrawable* gldrawable;

	glcontext = gtk_widget_get_gl_context(widget);
	gldrawable= gtk_widget_get_gl_drawable(widget);

	gdk_gl_drawable_gl_begin(gldrawable,glcontext);


	build_projection();

	initialize_gl_sence(displayrc);

	gdk_gl_drawable_gl_end(gldrawable);

}

static gboolean on_resize(GtkWidget *widget, GdkEventConfigure *event,gpointer user_data)
{
	GdkGLContext * glcontext;
	GdkGLDrawable* gldrawable;

	glcontext = gtk_widget_get_gl_context(widget);
	gldrawable= gtk_widget_get_gl_drawable(widget);

	if (gdk_gl_drawable_gl_begin(gldrawable, glcontext))
	{
		glViewport(0, 0, event->width, event->height);

		build_projection();

		gdk_gl_drawable_gl_end(gldrawable);
	}

	return FALSE;
}


src/drawing.c
/*
 * drawing.c
 *
 *  Created on: 2010-9-25
 *      Author: cai
 */


#include <config.h>

#include <stdlib.h>
#include <time.h>
#include <locale.h>
#include <glib.h>
#include <glib/gi18n.h>
#include <gtk/gtk.h>
#include <gtk/gtkgl.h>
#include <GL/glu.h>
#include <GL/glx.h>
#include <GL/glext.h>

#include "byllk.h"


GLubyte pixels[3*64*64];

void make2dImage(void)
{
	int j,i;
	for(i = 0 ; i < 64;i++)
	{
		for (j = 0; j < 64; j++)
		{
			pixels[i*j] = 255;
			pixels[i*j + 1 ] = 255-2*j;
			pixels[i*j + 2 ] =  255-2*i;
		}
	}
}

static void draw_quad(double xsize,double ysize)
{
	double	point[4][2] = { {0,0},{0,ysize},{xsize,ysize},{xsize,0}};

	glBegin(GL_QUADS);

	glVertex2dv(point[0]);
	glVertex2dv(point[1]);
	glVertex2dv(point[2]);
	glVertex2dv(point[3]);

	glEnd();
}

static void draw_quads()
{
	int i,j;

	float color[2][3] = { {.7,.7,.7},{1,1,1} };

	double xsize =2.0/LINEX;
	double ysize =2.0/LINEY;

	glPushMatrix();

	glTranslated(-1,-1,0); //移到左下角

	for (i = 0; i < 16; i++)
	{
		glPushMatrix(); //这样做是为了排除累计误差

		glTranslated(0,i*ysize,0);

		for (j = 0; j < 25; j++)
		{
			glColor3fv(color[ (i^ j )&1 ]);
			draw_quad(xsize,ysize);
			glTranslated(xsize,0,0);
		}
		glPopMatrix();
	}
	glPopMatrix();
}

static void draw_ball(double xsize,double ysize,float color[])
{
	GLfloat Notes[8]={0.0f,0.0f,0.0f,0.0f,1.0f,1.0f,1.0f,1.0f};

	GLfloat	controlpoints[4][4][3]={
			{{-0.06,1,-0.05},	{0.2,1,0},		{0.8,1,0},	{1.06,1,-0.05},},
			{{0,0.8,0},	{0.2,0.8,0.4},	{0.8,0.8,0.4},	{1,0.8,0},},
			{{0,0.2,0},	{0.2,0.2,0.4},	{0.8,0.2,0.4},	{1,0.2,0},},
			{{-0.06,0,-0.05},	{0.2,0,0},		{0.8,0,0.0},	{1.06,0,-0.05},},
	};


	glPushMatrix();

	glScaled(xsize,ysize,(xsize+ysize)/2);
	glTranslated(0.1,0.1,0);
	glScaled(0.8,0.8,1);

	glColor3fv(color);
	glMaterialfv(GL_FRONT_AND_BACK,GL_DIFFUSE,color);
//	glMaterialf(GL_FRONT_AND_BACK,GL_SHININESS,70);
//	glMaterialf(GL_FRONT_AND_BACK,GL_SPECULAR,1);

	//让我们制造有亮点的小块吧，(*^__^*) 嘻嘻……

	GLUnurbs *Nurb = gluNewNurbsRenderer();

	//设置NURBS曲面对象属性
	gluNurbsProperty(Nurb,GLU_SAMPLING_TOLERANCE,150.0f);
//	gluNurbsProperty(Nurb,GLU_DISPLAY_MODE,GLU_FILL);
	//开始NURB区间绘制
	gluBeginSurface(Nurb);
	//定义NURBS曲面的形状
	gluNurbsSurface(Nurb,8,Notes,8,Notes,4*3,3,&controlpoints[0][0][0],4,4,GL_MAP2_VERTEX_3);
	//结束曲面绘制
	gluEndSurface(Nurb);
	gluDeleteNurbsRenderer(Nurb);

	glPopMatrix();
}

static void make_display_list(DisplayResource * displayrc)
{
	int i;
	GLint	ball_list;

	float	color[COLOR_NUM][4] = {
			{1,0.1,0.1,1},{1,1,0.1,1},{0.1,1,0.1,1},{0.2,0.3,1,1},
			{1,0.2,1,1},{0,0,0,.5},{0.6,.6,0.6,1}
	};

	double xsize =2.0/LINEX;
	double ysize =2.0/LINEY;

	ball_list = glGenLists(COLOR_NUM);

	for(i=0;i<COLOR_NUM;i++)
	{
		displayrc->colorball_displaylist[i] = ball_list + i;

		glNewList(displayrc->colorball_displaylist[i],GL_COMPILE);

		draw_ball(xsize,ysize,color[i]);

		glEndList();
	}

	displayrc->background_displaylist = glGenLists(1);
	glNewList(displayrc->background_displaylist,GL_COMPILE);
	draw_quads();
	glEndList();

/*	displayrc->clicked_displaylist = glGenLists(1);
	glNewList(displayrc->clicked_displaylist,GL_COMPILE);
	glColor4f(0,0,0,0.5);
	draw_quad(xsize,ysize);
	glEndList();
*/
}

gboolean on_draw(GtkWidget * widget , GdkEventExpose * event, gpointer user_data)
{
	GdkGLContext * glcontext;
	GdkGLDrawable* gldrawable;
	DisplayResource	* displayrc;

	int x,y;

	double xsize =2.0/LINEX;
	double ysize =2.0/LINEY;

	float	color[COLOR_NUM][3] = {
			{1,0,0},{1,1,0},{0,1,0},{0,0,1}
	};

	g_object_get(widget,"user-data",&displayrc,NULL);

	glcontext = gtk_widget_get_gl_context(widget);
	gldrawable= gtk_widget_get_gl_drawable(widget);

	gdk_gl_drawable_gl_begin(gldrawable,glcontext);

	glLoadIdentity();

	glClear(GL_DEPTH_BUFFER_BIT);

	//这样背景就显示好了，(*^__^*) 嘻嘻……
	glCallList(displayrc->background_displaylist);

	glEnable(GL_LIGHTING);
	glEnable(GL_MULTISAMPLE);

	glPushMatrix();

	glTranslated(-1,-1,0); //移到左下角

	for (y= 0; y < LINEY; y++)
	{
		glPushMatrix();

		glTranslated(0,ysize*y,0);

		for (x = 0; x < LINEX; x++)
		{
			if(displayrc->balls[x][y]>=0)
			{
				glCallList(displayrc->colorball_displaylist[displayrc->balls[x][y]]);
			}
			glTranslated(xsize,0,0);
		}
		glPopMatrix();
	}

	glPopMatrix();

	glDisable(GL_LIGHTING);
	glDisable(GL_MULTISAMPLE);
/*
	//clicked box
	if (displayrc->clicked.x >= 0 && displayrc->clicked.y >= 0)
	{
		glPushMatrix();
		glTranslated(xsize * displayrc->clicked.x - 1, ysize* displayrc->clicked.y - 1, 0);
		glCallList(displayrc->clicked_displaylist);
		glPopMatrix();
	}
*/
	if(gdk_gl_drawable_is_double_buffered(gldrawable))
		gdk_gl_drawable_swap_buffers(gldrawable);
	else
		glFlush();

//	g_error("%s",gluErrorString(glGetError()));


	gdk_gl_drawable_gl_end(gldrawable);
	return TRUE;
}

void initialize_gl_sence(DisplayResource *displayrc)
{

	glDepthFunc(GL_LEQUAL);
	glEnable(GL_BLEND);
	glEnable(GL_DEPTH_TEST);
	glEnable(GL_AUTO_NORMAL);
	glEnable(GL_NORMALIZE);

	glEnable(GL_LIGHT0);
	glEnable(GL_LINE_SMOOTH);
	glEnable(GL_POLYGON_SMOOTH);

//	glShadeModel(GL_FLAT);

	GLfloat	light_position[4] =  {
			1,1,20,1
	};

//	glLightfv(GL_LIGHT0,GL_POSITION,light_position);

	glSampleCoverage(0.9,GL_TRUE);

	make_display_list(displayrc);

}
